[DEMO] Learning to learn with Quantum NN via Classical NN

stfnmangini opened the pull request:
#### General information

**Name**
Stefano Mangini (stfnmangini)

**Affiliation**
University of Pavia

--------------------------------------------------------------------------------

**Title:**  
QML Demo implementation of the paper "Learning to learn with quantum neural networks via classical neural networks" by Verdon et al.

**Summary:**  
Using Pennylane and Tensorflow, this demo uses a classical recurrent neural network to assist the optimization of variational quantum algorithms, as proposed by Verdon et al in their paper. 

**Relevant references:**  
- Paper: https://arxiv.org/abs/1907.05415
- My Github repo with code: https://github.com/stfnmangini/Learning2learn

**Possible Drawbacks:** 

**Related GitHub Issues:**

co9olguy commented:
Awesome, thanks for the contribution @stfnmangini!

josh146 commented:
Hey @stfnmangini, I made a small commit to fix the build. It should all be working now 🙂 Simply click 'details' for 'build artifact',

![image](https://user-images.githubusercontent.com/2959003/106131272-74119080-619d-11eb-8626-a041436291b0.png)

to browse the rendered website for this PR.

stfnmangini commented:
Hi @josh146, thank you very much! That's my first contribution so I'm a bit unsure about things to do 😅 . 
Hope I did all correctly. Btw, I had a look at the rendered website and it looks definitely good! :D

antalszava reviewed (commented):
@stfnmangini this is really great and exciting, thank you so much! 🥇

It seems that the overall the runtime is around 19 minutes for the demo, so would suggest converting it into a static demonstration. Overall this does not affect any parts of the demo.

This just means that rather than running the code on `readthedocs` to produce outputs, the results are pre-recorded in the script based on the output of a local run.

So the following needs to be done:
* Each statement in the script that prints an output has to be pre-recorded
    * see the [braket parallel demo](https://pennylane.ai/qml/demos/braket-parallel-gradients.html) and [this part of it](https://github.com/PennyLaneAI/qml/blob/3f1332949a5070f98848f01478cac22bdf105a06/demonstrations/braket-parallel-gradients.py#L388) for printed outputs
    * see the [Pytorch noisy demo](https://pennylane.ai/qml/demos/pytorch_noise.html) and [this part of it](https://github.com/PennyLaneAI/qml/blob/3f1332949a5070f98848f01478cac22bdf105a06/demonstrations/pytorch_noise.py#L159) for including pictures
* The script needs to be renamed to `learning2learn.py` (this is to tell `readthedocs` not to run the demonstrations, but just render the content of the file).

stfnmangini commented:
Hi @antalszava 👋 
Thanks for all your suggestions! ❤️ 
I implemented them and also converted the demo to be a static one, thus adding outputs and figures, and modifying the files accordingly. It seems everything is working correctly, hope it is actually so!

glassnotes commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+r"""
+Learning to Learn with Quantum NN via Classical NN
+==================================================
+*Author: Stefano Mangini (mangini.stfn@gmail.com)*
```
```suggestion
*Author: Stefano Mangini (mangini.stfn@gmail.com). Posted: XX Feb 2021. Last updated: XX Feb 2021.*
```
@stfnmangini When everything's ready to go, please modify the two dates here to the merge date :+1: Thanks!

stfnmangini replied:
Will do! Thanks :)

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+   tutorial_qaoa_maxcut QAOA for MaxCut problem
+
+
+In this tutorial we recreate the architecture proposed by Verdon et
```
```suggestion
In this demo we recreate the architecture proposed by Verdon et
```
This is the current style: refer to them as demos :-)

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+
+In this tutorial we recreate the architecture proposed by Verdon et
+al. in the paper *Learning to learn with quantum neural networks via
```
```suggestion
al. in *Learning to learn with quantum neural networks via
```
to avoid repeated use of "paper"

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+In this paper, classical recurrent neural networks are used to assist
+the optimization of variational quantum algorithms.
+
+We will start with a brief theoretical overview explaining the problem
```
Purely stylistic suggestion: I personally think text reads better when using "we" instead of "we will". For example, here you could write
"we start with a brief..."
"After that, we take a deep dive into..."

stfnmangini replied:
Definitely agree! Thanks for pointing this out :D

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+We will start with a brief theoretical overview explaining the problem
+and the setup used to solve it. After that, we will deep dive into the
+code to build a fully functioning model, ready to be further developed
+or customized for your own needs. Without further ado, let’s start!
```
```suggestion
or customized for your own needs. Without further ado, let’s begin!
```
To avoid repeated use of "start"

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+------------------------------------------------------------------
+
+Recently, a big effort by the quantum computing community has been
+devoted to the study of variational quantum algorithms (VQAs, for short)
```
```suggestion
devoted to the study of variational quantum algorithms (VQAs)
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+Recently, a big effort by the quantum computing community has been
+devoted to the study of variational quantum algorithms (VQAs, for short)
+which leverage quantum circuits with fixed shape but with tunable
```
```suggestion
which leverage quantum circuits with fixed shape and tunable
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+Recently, a big effort by the quantum computing community has been
+devoted to the study of variational quantum algorithms (VQAs, for short)
+which leverage quantum circuits with fixed shape but with tunable
+parameters to solve a desired target task. The idea is similar to
```
```suggestion
parameters to solve a desired task. The idea is similar to
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+which leverage quantum circuits with fixed shape but with tunable
+parameters to solve a desired target task. The idea is similar to
+classical neural networks, where the weights of the network are
+optimized during training to perform best in regression or
```
```suggestion
optimized during training to improve performance in regression or
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+the particular task at hand — its tunable parameters are optimized
+iteratively by minimizing a cost (or loss) function, which measures how
+good the quantum algorithm is performing (see [#vqas]_ for a
+thorough overview on VQAs).
```
Great paragraph! :star_struck:

stfnmangini replied:
Super glad you like it! Awesome! 🤩

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+thorough overview on VQAs).
+
+A major challenge for VQAs relates to the optimization of tunable
+parameters, which was shown to be a very hard task to perform (a
+possible challenge is posed by the so called *barren plateaus*). In
```
Consider removing the sentence in parenthesis and instead add one or two key citations

stfnmangini replied:
That makes sense, will do.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+particular, parameter initialization plays a key role in this scenario,
+since initializing the parameters in the proximity of an optimal
+solution leads to faster convergence and better results with respect to
+random initialization, which could bring the model in a region in
+parameter space far from any good solution. Thus, a good initialization
```
```suggestion
random initialization. Thus, a good initialization
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+random initialization, which could bring the model in a region in
+parameter space far from any good solution. Thus, a good initialization
+strategy is crucial to promote the convergence of local optimizers to
+local extrema and to select reasonably good local minima. Here, by local
```
```suggestion
local extrema and to select reasonably good local minima. By local
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+local extrema and to select reasonably good local minima. Here, by local
+optimizer, we mean a procedure that moves from one solution to another
+by small (local) changes in parameter space. These are opposed to global
+search methods, which take into account the whole (or at least a large
+section) parameter space to propose a new solution.
```
```suggestion
search methods, which take into account large sections of
parameter space to propose a new solution.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+One such strategy could come from the classical machine learning
+literature.
+
+Solution: (Classical) Recurrent Neural Networks
```
```suggestion
Solution: Classical Recurrent Neural Networks
```
Or simply remove "classical"

stfnmangini replied:
Mmhh...Yeah, that "Classical" may be redundant. In my mind, it was used to stress that this is not a quantum generalization of a RNN, but just a purely classical one applied to a quantum problem.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+variational quantum algorithms, as shown in the figure below. The cost
+function used is the expectation value :math:`\langle H \rangle_{\boldsymbol{\theta}} = \langle \psi_{\boldsymbol{\theta}} | H | \psi_{\boldsymbol{\theta}}\rangle`
+of a Hamiltonian :math:`H` with respect to the parametrized state
+:math:`|\psi_\boldsymbol{\theta}\rangle` evolved by applying the variational quantum circuit to the zero state.
```
maybe add a short mathematical description of the "zero state", e.g., |00..0>

stfnmangini replied:
Yeah, that's needed! Thanks :)

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+Given parameters :math:`\boldsymbol{\theta}_{t-1}` of the variational quantum circuit,
+the cost function :math:`y_{t-1}`, and the hidden state of the
+classical network :math:`\boldsymbol{h}_{t-1}` at the previous time step, the
+classical computer (the Recurrent Neural Network) proposes a new
```
```suggestion
recurrent neural network proposes a new
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+classical network :math:`\boldsymbol{h}_{t-1}` at the previous time step, the
+classical computer (the Recurrent Neural Network) proposes a new
+guess for the parameters :math:`\boldsymbol{\theta}_t`, which are
+then fed into the quantum computer (the QAOA circuit) to evaluate the
```
At this stage you haven't really mentioned that the quantum computer is implementing a QAOA circuit. Maybe remove "(the QAOA circuit)". 

In general, I recommend avoiding excessive use of parenthesis in text. It's something scientists do often, but great writers avoid :wink:

stfnmangini replied:
Oh, this reminds me of all the times my High School Italian teacher complained about my writing style 🤣 
You're definitely right, thanks for this suggestion :)

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+step :math:`\boldsymbol{h}_t`. The RNN itself has trainable parameters :math:`\phi`,
+and hence it applies the parametrized mapping:
+
+.. math::  \boldsymbol{h}_{t+1}, \boldsymbol{\theta}_{t+1} = \text{RNN}_{\phi}(\boldsymbol{h}_{t}, \boldsymbol{\theta}_{t}, y_{t})
```
```suggestion
.. math::  \boldsymbol{h}_{t+1}, \boldsymbol{\theta}_{t+1} = \text{RNN}_{\phi}(\boldsymbol{h}_{t}, \boldsymbol{\theta}_{t}, y_{t}),
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+which generates a new suggestion for the variational parameters as well
+as a new internal state. Upon training the weights :math:`\phi`, the RNN
+eventually learns a good heuristic to suggest optimal parameters for the
+quantum circuit.
```
I may need to re-read, but after a first pass, it's not clear in what sense the RNN is doing something different than just optimizing the variational parameters :thinking: 

So is it correct to view this as simply a different optimization strategy? :thinking: If so, I'm not sure in what sense we're "learning to learn". Still cool!

stfnmangini replied:
The idea is that we are not directly optimizing the variational parameters of the quantum circuit, but instead we let the RNN figure out how to do that. In this sense, we are learning (training the RNN) how to learn (how to optimize a variational quantum circuit).  
Hope this makes more sense 😅

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+There are multiple VQAs for which this hybrid training routine could
+be used, some of them directly analyzed in [#l2l]_. In the
+following, we will only focus on one of such examples, that is the use
+of the QAOA (Quantum Approximate Optimization Algorithm) for solving
+the MaxCut problem for graphs. Thus, referring to the picture above,
```
```suggestion
following, we focus on one such example, the
Quantum Approximate Optimization Algorithm (QAOA) for solving
the MaxCut problem for graphs. Thus, referring to the picture above,
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+the MaxCut problem for graphs. Thus, referring to the picture above,
+the shape of the variational circuit is the one dictated by the QAOA
+ansatz, and such a quantum circuit is used to evaluate the cost
+Hamiltonian :math:`H` of the MaxCut problem of some graphs.
```
```suggestion
Hamiltonian :math:`H` of the MaxCut problem.
```
And add a link to a site explaining MaxCut, e.g., wikipedia

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+the shape of the variational circuit is the one dictated by the QAOA
+ansatz, and such a quantum circuit is used to evaluate the cost
+Hamiltonian :math:`H` of the MaxCut problem of some graphs.
+If you wish to refresh your memories, check out this great tutorial on
```
:star_struck: Thanks for the shout-out!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+how to use QAOA for solving graph problems: https://pennylane.ai/qml/demos/tutorial_qaoa_intro.html
+
+.. note::
+   Running the whole tutorial requires approx. ~10m.
```
Not sure if including this is a good idea :thinking:

stfnmangini replied:
Maybe we could stick to PennyLane's style. Do other demos have that?

josh146 replied:
I don't think other demos include it, tagging @glassnotes for her thoughts. I'm not against it though, it might be useful to include in case someone downloads it and wonders if it is working correctly?

glassnotes replied:
If it's only 10 minutes, why not just make it executable? The runtime will be at the bottom then.

If it can't be made executable I'm in favour of keeping the note, so that users know what to expect. (Some of the other demos say something to the effect of "this takes some time" but are not explicit, and the runtimes are like 30 minutes!) I'd rather be explicit :+1:

stfnmangini replied:
Actually, it takes ~10min (maybe a bit more) to execute on my pc. But when I opened this PR the script took ~20min to run on CircleCI. 
Since it took too long, I made it pre-rendered. 

Maybe I could leave the script pre-rendered and change the note to "~20min" (or get rid of the note at all)?

glassnotes replied:
Ah gotcha, yes 20 mins is a bit long for executable. I'd change the note to say something like "up to 20 minutes", and leave it as pre-rendered then.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# seamlessly with **TensorFlow**, which will be used for creating the RNN.
+#
+
+# Quantum Machine Learning
```
Amazing idea to provide comments for imports! :exploding_head: 

Somehow we never did this before. I really like it!

stfnmangini replied:
It seemed like a nice thing to do!
Yaaay 🎉

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+import matplotlib.pyplot as plt
+import random
+
+# Set the seed for reproducibility of results
```
```suggestion
# Fix the seed for reproducibility, which affect all random functions in this demo
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+import random
+
+# Set the seed for reproducibility of results
+# These seeds affect all random functions in this tutorial
```
```suggestion
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# Generation of training data: graphs
+# -----------------------------------
+#
+# As for any Machine Learning project, the first step is to gather (or in
+# our case create) a good dataset, which will be used to train the model
```
```suggestion
# As for any machine learning project, the first step is to gather or
# create a good dataset that will be used to train the model
```

stfnmangini replied:
Damn parenthesis!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# our case create) a good dataset, which will be used to train the model
+# and test its performances. In our case, we are analyzing the MaxCut
+# problem, which deals with the problem of finding a good binary partition
+# of nodes in a *graph*, such that the number of edges *cut* by such
```
```suggestion
# of nodes in a graph, such that the number of edges *cut* by such
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# separation is maximized. For this reason, we start by generating some
+# random graphs :math:`G_{n,p}` where:
+#
+# * :math:`n` is the number of nodes in the graphs,
```
```suggestion
# * :math:`n` is the number of nodes in each graph,
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+#
+
+
+def generate_graphs(n_graphs, n_nodes, p_edge):
```
Nice!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# Now that we have a dataset, we move on by creating the QAOA quantum
+# circuits using PennyLane’s built-in sub packages. In particular, using
+# PennyLane’s ``qaoa`` module, we will able to create fully functioning
+# quantum circuits for the MaxCut problem, with very few lines of code.
```
:heart_eyes:

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+    """Creates a cost function by encoding a Hamiltonian that solves the MaxCut problem using QAOA.
+
+    This function uses nested functions for the creation of the quantum circuit. When called,
+    it instantiates the QAOA circuit for the input graph and outputs a QAOA cost function.
+    Such function, given some parameters, evaluates the MaxCut cost function for the given graph.
```
```suggestion
    it instantiates the QAOA circuit for the input graph and outputs a QAOA cost function 
    that evaluates the MaxCut cost function for the given graph.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+        return cost(params)
+
+    return hamiltonian
```
I wonder if you actually need to wrap all these functions into a single one instead of using them separately. Usually keeping functions separated helps with flexibility and debugging

stfnmangini replied:
This was to have all the QAOA's stuff self-contained. Like a black-box that you should not tweak with.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+
+######################################################################
+# Before moving on, let’s see how to use these functions.
```
```suggestion
# Before continuing, let’s see how to use these functions.
```
"moving on" kind of implies "forgetting about"

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+######################################################################
+# Now we optimize the parameters to reduce the cost function ``cost``. We
+# can do this using ``tf.GradientTape()``, which works directly thanks to
+# PennyLane’s seamless integration with TensorFlow when using the
```
```suggestion
# PennyLane’s integration with TensorFlow when using the
```
I know it is indeed seamless, but let's not repeat that too much :grin:

stfnmangini replied:
Ahahaha seems legit! 😄

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# PennyLane’s seamless integration with TensorFlow when using the
+# ``default.qubit.tf`` device.
+#
+# .. note::
```
Nice!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# As you can see, the loss function is correctly minimized throughout the
+# optimization process, which eventually yields an optimal value for the
+# angles in the QAOA ansatz.
+#
```
Note to myself: stopping review here

ixfoduap reviewed (commented):
Hi @stfnmangini,

The demo looks great! Really excited to see these external contributions coming in. I only got up to line 370, but will continue my review later this week.

I have no major comments, only a collection of suggestions for further improving readability. The demo is in great shape, I'm just trying my best to improve it even further :rocket:

antalszava reviewed (approved):
Looks good to me! :100: This is really amazing @stfnmangini :blush: 

From my side, this is good to be merged, but let's do that once the other comments and suggestions have also been addressed and reviews finalized.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# ------------------------------
+#
+# So far, we have defined the machinery which lets us build the QAOA
+# algorithm for solving the MaxCut problem of a graph.
```
```suggestion
# algorithm for solving the MaxCut problem.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# algorithm for solving the MaxCut problem of a graph.
+# Now we wish to implement the Recurrent Neural Network architecture
+# explained previously. In particular, as proposed in the original
+# paper, we will build a custom model of an LSTM (Long-Short Term
+# Memory) network, capable of handling the hybrid data passing between
```
```suggestion
# paper, we will build a custom model of an Long-Short Term
# Memory (LSTM) network, capable of handling the hybrid data passing between
```
Normally the full name comes first, then you introduce an acronym for simplicity, not the other way around :wink:

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+######################################################################
+# First of all, let’s define the elemental building block of the model,
+# that is an LSTM cell (see `TensorFlow
```
```suggestion
# an LSTM cell (see `TensorFlow
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+#
+
+# Number of layers in QAOA ansatz. The higher the better in terms of performance,
+# but it also gets more computationally hard. For simplicity, we will stick to the single layer case.
```
```suggestion
# but it also gets more computationally expensive. For simplicity, we will stick to the single layer case.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+n_layers = 1
+
+# Define a single LSTM cell. The cell has two units per layer
+# since each layer in the QAOA ansatz make use of two angles (parameters).
```
```suggestion
# since each layer in the QAOA ansatz makes use of two parameters.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+######################################################################
+# At this stage, we seek to reproduce the recurrent behavior depicted in
+# the picture above, outlining the functioning of RNN as a black-box
```
```suggestion
# the picture above, outlining the functioning of an RNN as a black-box
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# the picture above, outlining the functioning of RNN as a black-box
+# optimizer. We do so by defining two functions:
+#
+# * ``rnn_iteration``: accounts for the computations happening on a single time step in the figure,
+#                      that is it performs the calculation inside the CPU and evaluates the quantum circuit on the QPU to obtain
```
```suggestion
# * ``rnn_iteration``: accounts for the computations happening on a single time step in the figure,
#                      that is, it performs the calculation inside the CPU and evaluates the quantum circuit on the QPU to obtain
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+def rnn_iteration(inputs, graph_cost, n_layers=1):
+    """Perform a single time step in the computational graph of the custom RNN.
+
+    The function takes as inputs, the outputs from a previous time step and a graph cost function,
```
```suggestion
    The function takes as inputs the outputs from a previous time step and a graph cost function,
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# The cost function
+# ~~~~~~~~~~~~~~~~~~~~~~
+#
+# A key part in the ``recurrent_loop`` function, is given by the
```
```suggestion
# A key part in the ``recurrent_loop`` function is given by the
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# *observed improvement*, for simplicity here we use an easier cost
+# function :math:`\cal{L}(\phi)` defined as:
+#
+# .. math:: \cal{L}(\phi) = {\bf w} \cdot {\bf y}_t(\phi)
```
```suggestion
# .. math:: \cal{L}(\phi) = {\bf w} \cdot {\bf y}_t(\phi),
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# ~~~~~~~~~~~~~~~~~~~~~~
+#
+# Now all the cards are on the table and we just need to prepare a
+# training routine and then run it!
```
Great sentence!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# .. note::
+#     For a QAOA ansatz using one single layer, and for a training set of
+#     20 graphs, each epoch takes approximately ~1m to run on a standard
+#     laptop. Be careful when using bigger dataset or training for larger
```
```suggestion
#     laptop. Be careful when using bigger datasets or training for larger
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# Apply the RNN (be sure that training was performed)
+res = recurrent_loop(new_cost, intermediate_steps=True)
+
+# Extract all angle suggestions
```
Oh! I had to come to this point to understand what the main point of this strategy is. The idea is that by training on a dataset of graphs, the RNN can subsequently be used to provide suggestions for starting points on new graphs! That's cool!

But then it would be best to explain this clearly and succinctly from the start!

stfnmangini replied:
Yes that's the idea! I'll add some more sentences at the start and try to be more clear about the main idea :)

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# Comparison with standard Stochastic Gradient Descent (SGD)
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# That’s remarkable! The RNN learned to propose new parameters such that
```
Great paragraph!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# Gradient Descent). Thus, as the authors suggest, the (trained) RNN can
+# be used for few iterations at the start of the training procedure to
+# initialize the parameters of the quantum circuit close to an optimal
+# solution. Then, a standard optimizer like the SGD, can be used to
```
```suggestion
# solution. Then, a standard optimizer like the SGD can be used to
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# fine-tune the proposed parameters and reach even better solutions.
+# While on this small scale example the benefits of using an LSTM to
+# initialize parameters may seem modest, on more complicated instances
+# and problems it can make a big difference, since, on random
```
```suggestion
# and problems it can make a big difference, since on random
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+
+
+######################################################################
+# Loss landscape in parameter space
```
I've reached this point after a busy day at work and I don't have the energy to read! The tutorial is great, but sometimes trying to do too much is not optimal. I would consider removing these sections and jumping to the final remarks

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+#    ``rnn_iteration`` and ``recurrent_loop`` to actual ``Keras Layers``
+#    and ``Models``. This way, by compiling the model before the training
+#    takes place, ``TensorFlow`` can create the computational graph of the
+#    model and train more efficiently. If you’re interested, you can find
+#    some ideas below to start working on it.
```
```suggestion
#    model and train more efficiently.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+# hesitate to contact me, I’ll be very happy to help you as much as I can
+# 😁
+#
+# Have a great quantum day!
```
Great place to end!

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1205 @@
+#
+
+
+######################################################################
```
In line with my comment above, it's a lot to ask for the reader to actually go through this material. I would suggest omitting it from the demo

josh146 replied:
Curious fly-by question: would an 'appendix' work? It's not something we've tried yet in other demos, but might be a nice way of keeping the generating code available for interested users without obscuring the story.

Another approach could be splitting this into _two_ demos! That something we did with the variational linear solver demo, since we noticed it could be neatly divided into two sections, each one with a slightly different focus, but still part of the same overall story:

![image](https://user-images.githubusercontent.com/2959003/106702914-20cb9200-6624-11eb-8fab-bc36bd20bdb7.png)

stfnmangini replied:
Hi @ixfoduap, thanks very much for your thorough review! :)
You definitely improved the quality and readability of the demo 😄 🎉 

As for the length, we could omit the section "Loss landscape in parameter space" without harm, since it is not essential to the demo. Though, I feel the "Generalization performances" section to be quite important.

Alternatively, as @josh146 suggested, maybe it could be useful to move these to an Appendix? Besides, I think that dividing it into two demos would not be ideal, because I cannot find a neat distinction in two "standalone" topics. 

Thanks again for your suggestions! :D

antalszava replied:
Hi @stfnmangini, chatted with @ixfoduap and an appendix could be a nice solution. :)

Feel free to leave a note here or send a ping to discuss the details if any questions come up regarding the specifics.

stfnmangini replied:
Hi @antalszava! 
Thanks :D

I just pushed some changes, in particular:
1) Added a reference to the barren plateau issue;
2) Added a sentence at the start explaining more in detail the idea of the RNN approach;
3) Deleted the docstrings in function to free up some space;
4) Created an Appendix, and moved there auxiliary (but interesting) sections.

Now the core of the tutorial consists of 750 lines, hope that's good enough! 😅

ixfoduap reviewed (commented):
Excellent tutorial! :fire: I was very glad to see that you've made an effort to replicate our style :tada: 

Besides all the minor suggestions I made, I have two major comments:
- The tutorial is really, really long! This is not ideal as very few users will have the stamina to go through the entire thing. In my comments, I suggested some sections that could be removed outright. However, I think the demo would benefit greatly from being much more concise. *Can you bring the entire demo to under 500 lines?* That would be ideal
- I think that the main motivation for this technique should be clearer in the beginning: our goal is to train a neural network on a dataset of problems to make better suggestions for initial circuit parameters on new problems. This leads to increased efficiency in the optimization and better solutions!

I'm asking for extra work, but trust me, it will be well worth it!

ixfoduap reviewed (commented):
Hi @stfnmangini,

This looks great! I still need some more time to take a closer look at your changes, but it already looks like a significantly improved version with more manageable length. 

Great job! :muscle:

antalszava commented on demonstrations/learning2learn.py:
```diff
@@ -115,7 +123,7 @@
 
 .. note::
-   Running the whole tutorial requires approx. ~10m.
+   Running the whole tutorial requires approx. ~15m.
```
Would this be a number when run including the `Appendix`? Maybe it's better to have a separate note there.

stfnmangini replied:
On my pc, the whole notebook (including Appendix) required about ~10-15min to run. 
Here, it took 20min to run the whole (including Appendix) script. 
In the note, I wrote ~15min because that was an average between the runtime on my computer and on CircleCI.

However, now that we have created an Appendix, I expect the main part of the script (without Appendix) to require roughly ~7/8min on my pc, and 12/13min on CircleCI). So, what do you suggest? 
I think that writing explicitly "13min without the appendix" should be fine.

stfnmangini replied:
Thanks @antalszava! You are right, the toc was maybe too long. 

I eliminated all subsections and substituted them with **bold** text so that they are still visible while reading, without occupying space in the toc. 

I just left subsections in the Appendix where they are needed since each part is actually independent from the others.
Is this ok?

antalszava replied:
> I think that writing explicitly "13min without the appendix" should be fine.

:+1: My comment more so referred to separating runtime without the Appendix at this point.

> I eliminated all subsections and substituted them with bold text so that they are still visible while reading, without occupying space in the toc.

Oh nice, this is a very neat solution!! :100: :blush: 

> I just left subsections in the Appendix where they are needed since each part is actually independent from the others.
Is this ok?

:100: TOC is looking great :+1:

stfnmangini replied:
Thanks, @antalszava! :D 
Now I'll brag around with this new Best TOC evah.

antalszava reviewed (commented):
@stfnmangini thank you for cutting down on the main part of the tutorial!

In the rendered version of the tutorial, as the `Contents` section now on the right has grown (e.g., with an `Appendix`), I'd suggest removing/merging some headers (e.g., `VQAs in focus: QAOA for MaxCut` (QAOA is mentioned in a later header), `Importing the required packages`, `The cost function`, `Training`, etc.). This will help readers notice that there is still content after the tutorial, should they glimpse on the right sidebar.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+plt.legend()
+ax.set_xticks([0, 5, 10, 15, 20]);
+
+######################################################################
```
It's a strange plot. Why does the x-axis go to so many iterations if the curve is only plotted for ~5? Why is the curve dotted (-.)? 

I know you mention this in the text below, but I feel that the fact that a minimum is observed after just a single step deserves a short explanation since this behaviour is atypical.

stfnmangini replied:
It is dotted in accordance with the next picture, as you noticed. Will restrict the range in [0,15].

As for the explanation, I'll try to add a sentence on that.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+######################################################################
+# *Hurray!* 🎉🎉
+#
+# As is it clear from the picture, the RNN reaches a better minimum in
+# fewer (but comparable) iterations than the standard SGD (Stochastic
+# Gradient Descent). Thus, as the authors suggest, the (trained) RNN can
+# be used for few iterations at the start of the training procedure to
+# initialize the parameters of the quantum circuit close to an optimal
+# solution. Then, a standard optimizer like the SGD can be used to
+# fine-tune the proposed parameters and reach even better solutions.
+# While on this small scale example the benefits of using an LSTM to
+# initialize parameters may seem modest, on more complicated instances
+# and problems it can make a big difference, since on random
+# initialization of the parameters, standard local optimizer may
+# encounter problems finding a good minimization direction (for further
+# details, see [#l2l]_, [#vqas]_).
```
```suggestion
# As is it clear from the picture, the RNN reaches a better minimum in
# fewer iterations than the standard SGD. 
Thus, as the authors suggest, the trained RNN can
# be used for a few iterations at the start of the training procedure to
# initialize the parameters of the quantum circuit close to an optimal
# solution. Then, a standard optimizer like the SGD can be used to
# fine-tune the proposed parameters and reach even better solutions.
# While on this small scale example the benefits of using an LSTM to
# initialize parameters may seem modest, on more complicated instances
# and problems it can make a big difference, since on random
# initialization of the parameters, standard local optimizer may
# encounter problems finding a good minimization direction (for further
# details, see [#l2l]_, [#vqas]_).
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+# Final remarks
+# -----------------
+#
+# In this tutorial we saw how to use a recurrent neural network, an LSTM
+# in particular, as a black-box optimizer to initialize the parameters in
+# a variational quantum circuit close to an optimal solution. In order to
+# do that, we connected MaxCut QAOA quantum circuits built in PennyLane
+# with an LSTM built with TensorFlow, and we used a custom hybrid training
+# routine to optimize the whole network.
```
```suggestion
# In this demo we saw how to use a recurrent neural network
# as a black-box optimizer to initialize the parameters in
# a variational quantum circuit close to an optimal solution. 
# We connected MaxCut QAOA quantum circuits in PennyLane
# with an LSTM built with TensorFlow, and we used a custom hybrid training
# routine to optimize the whole network.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+#
+# Such architecture proved itself to be a good candidate for the
+# initialization problem of Variational Quantum Algorithms, since it
+# reaches good optimal solutions in very few iterations. Besides, the
+# architecture is quite general, since the same machinery can be used for
```
```suggestion
# reaches good optimal solutions in very few iterations. Besides, the
# architecture is quite general since the same machinery can be used for
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+#    ``rnn_iteration`` and ``recurrent_loop`` to actual ``Keras Layers``
+#    and ``Models``. This way, by compiling the model before the training
+#    takes place, ``TensorFlow`` can create the computational graph of the
+#    model and train more efficiently. If you’re interested, you can find
+#    some ideas below to start working on it.
```
```suggestion
#    model and train more efficiently. You can find
#    some ideas below to start working on it.
```

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+plt.xlabel("Iteration", fontsize=12)
+ax.set_xticks([0, 5, 10, 15, 20]);
+
+######################################################################
```
Oh OK, the previous plot makes more sense once you see this one. Still I would recommend setting the x-axis from 0 to 15 only

stfnmangini replied:
Will do.

ixfoduap commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+# hesitate to contact me, I’ll be very happy to help you as much as I can
+# 😁
+#
+# Have a great quantum day!
```
You too! :wink:

stfnmangini replied:
Thanks, @ixfoduap! 
You again much improved the tutorial! I'll address your suggestions in the following days and make the changes!
:)

ixfoduap reviewed (approved):
Amazing work @stfnmangini !

If you address all my comments and suggestions, I'm happy to see the demo merged

stfnmangini commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+r"""
+Learning to Learn with Quantum NN via Classical NN
+==================================================
```
I agree. 
I used that one because it's the original name of the paper, but I like your restyled version.

stfnmangini commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+#
+
+
+def qaoa_from_graph(graph, n_layers=1):
```
Yes, that's a good thing to do :)

stfnmangini commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+# the picture above, outlining the functioning of an RNN as a black-box
+# optimizer. We do so by defining two functions:
+#
+# * ``rnn_iteration``: accounts for the computations happening on a single time step in the figure,
```
Ok, will have a deeper look.

stfnmangini commented on demonstrations/learning2learn.py:
```diff
@@ -0,0 +1,1101 @@
+# multiple times (epochs).
+#
+# .. note::
+#     For a QAOA ansatz using one single layer, and for a training set of
+#     20 graphs, each epoch takes approximately ~1m to run on a standard
+#     laptop. Be careful when using bigger datasets or training for larger
+#     epochs.
```
Seems reasonable! I'm changing that.

stfnmangini commented:
Ok @ixfoduap, I made some changes following your suggestions, in particular:
1) shortened lines that required lateral scrolling to read;
2) adjusted the range in the figures in [0, 15];
3) added a line explaining why the LSTM reaches a minimum in just one iteration;
4) addressed all your other suggestions.

The only thing left is to update the date on top of the demo when this will be definitely approved. 
Thanks so much :)

co9olguy commented:
Thanks for all your work on this @stfnmangini! And nice seeing you in the chat at QHack last week :)

ixfoduap reviewed (commented):
Good to go from my side! :shipit: :rocket: 

Congrats @stfnmangini, fantastic work

Final recommendation: take another full read yourself to make sure there aren't any typos and you're fully happy with the end result

stfnmangini commented:
> Thanks for all your work on this @stfnmangini! And nice seeing you in the audience at QHack last week :)

It was a fun project indeed, and I'm happy to contribute! As for Qhack, it was truly (and that's not just a saying) my pleasure. Really enjoyed it @co9olguy  🚀 👏  .

> Good to go from my side! :shipit: 🚀
> 
> Congrats @stfnmangini, fantastic work
> 
> Final recommendation: take another full read yourself to make sure there aren't any typos and you're fully happy with the end result

Thanks so much @ixfoduap ! Good suggestion: I'll give it another full read on Thursday, to spot any surviving typos!

glassnotes commented:
@stfnmangini Thanks for putting together this demo :grin: We usually post new demos on Tuesdays, so we're planning to do the formal release next week on 2 March (i.e., it will be announced publicly on Twitter that day). After you've finished your final pass, go ahead and update both dates in the byline before merging in (the best time to merge it will be Monday 1 March so that it's ready for the following day). :rocket:

stfnmangini commented:
Hi everyone! 
Just wanted to say that I went through another review round and found some surviving typos, which I corrected on the last commit of few days ago! As the date in the byline, I wrote 2 March. 
Hope everything is safe and sound now! :)

antalszava commented:
Thanks @stfnmangini! Merging this 🎉
